### PR TITLE
Add DynamicRegistryRegistry and DynamicRegistryFinalizeCallback

### LINKED ISSUE #1012: Hooks for adding custom registries into DynamicRegistryManager
Currently, the list of registry infos in `DynamicRegistryManager` is hardcoded; it would be nice to be able to add our own entries to that list.

### LINKED ISSUE #2111: DynamicRegistryTailCallback
I want to pr a callback that runs after the dynamic registry is set up, and gives you access to the registry manager like DynamicRegistrySetupCallback does.

If anyone has any better names, feel free to comment them.

### PR DESCRIPTION
**Pull request by ghost**
PR overview:

* `DynamicRegistryRegistry` : A helper class for registering new dynamic registries. (Name debatable)
* `DynamicRegistryFinalizeCallback` : An event fired just before DRM is immutablized.
* Updated test mod, moved dynamic registry related stuff to it's own initializer.

Closes #1012 
Closes #2111
Replaces #2112

### CODE REVIEW

**Comment by ghost** [comment]
By the way `impl.registry.sync.packet.NbtRegistryPacketHandler` seems to be marked for removal (has a "TODO: Remove"), I can remove it as a part of this pr, along with it's uses in test mod, which seems to be the only place it's used.

**Comment by ghost** [comment]
One last thing, subpackages for the api part of the module are inconsistent. It goes like `api.event.registry` and contains everything,  even though some are not event related. I didn't touch anything because it would be a breaking change, and I followed it. Just wanted to point it out.

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java
Comment (haykam821) [inline fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java]: This line is a bit unwieldy.

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java
Comment (haykam821) [inline fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java]: Is there a simpler way to do this, perhaps?

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java
Comment (haykam821) [inline fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java:52]: Should we mark this API/class as experimental?

**Review by haykam821 — CHANGES_REQUESTED** [review]
I personally think that a new module should be created for dynamic registries and possibly other registries, as was done in #1031.

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java
Comment (ghost) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java]: I can redirect the `List.of` and return a mutable list, though it might be compromising since modders would have full access to the list.

**Comment by ghost** [comment]
> I personally think that a new module should be created for dynamic registries and possibly other registries, as was done in #1031.

I partially agree, it doesn't fit any other module, but I also think it's too small to be a module. Expanding registry-sync module to a registry-related module seems more reasonable to me.

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java
Comment (ghost) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java]: Had to check it somehow, most I can do is something like below, which only makes it look slightly cleaner.
```java
String path = entry.key().getValue().getPath();
RegistryLoader.DYNAMIC_REGISTRIES.stream().filter(e -> e.key().getValue().getPath().equals(path)).findFirst().ifPresent(e -> {
```

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java
Comment (ghost) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java]: I decided to keep a private mutable list in the class.

**Comment by Technici4n** [comment]
@haykam821 I disagree. This:
- is fine in the regsync module, and
- should not be done using entrypoints.

**Comment by haykam821** [comment]
I was mainly referring to the module and not the API design. The `fabric-registry-sync-v0` module is getting overstuffed and should have been primarily a helper module rather than an API module.

**Comment by Technici4n** [comment]
It should simply contain all registry-related things, even though this isn't obvious due to the legacy name.

**Comment by Technici4n** [comment]
This module already has the DynRegSetupCallback, and various registry events that work with dynamic registries. Might as well support adding dynregs in there.

**Comment by haykam821** [comment]
The location of `DynamicRegistrySetupCallback` was a mistake when it was first merged, and it remains a mistake today. [My stance on this issue has stayed the same.](https://github.com/FabricMC/fabric/pull/2112#issuecomment-1147041087)

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java
Comment (haykam821) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryRegistry.java]: I think a stream is the wrong tool for this job, because filtering a stream doesn't short-circuit. If a stream is used, though, its calls should be formatted over several lines.

**Comment by ghost** [comment]
I'm converting this pr to draft because I just realized previously loaded dynamic registries can be accessed from the registry being loaded at the time. So I want to add methods to register before and after specific entries.

**Comment by ghost** [comment]
Okay I updated the pr, and also marked it as experimental like suggested since Mojang can make any changes to these during snapshots and pre-releases. I think it should stay as an experimental feature until full 1.19.3 release.

**Comment by ghost** [comment]
I need some opinions on this. Are return values for the registering methods really necessary? I can't think of any situation where people would need the entry.

**Comment by ghost** [comment]
Closes #1012

**Comment by ghost** [comment]
I don't want to overpopulate this pr but is a `DynamicRegistryFinalizeCallback` suitable for this pr. Or should I open another one after this. The idea is that it will be fired at the end of `RegistryLoader.load` just before DRM becomes immutable.

And which one is the best for the callback parameter:
* Directly pass the mutable DRM and hope no one calls `toImmutable()`.
* Pass a `DynamicRegistryView` and warn modders that they shouldn't use entry added/removed callback helper methods.
* Wrap the mutable DRM with a custom DRM that doesn't implement `toImmutable()`.

**Comment by ghost** [comment]
Hopefully updated this pr for the last time, at least I've got nothing else stuck on my mind.
I'll just be waiting for reviews from now on.

**Comment by dhyces** [comment]
looking forward to this

**Pull request closed without merging**
